### PR TITLE
Add "Show outliers" option

### PR DESCRIPTION
**Pull request by openjck**

### CODE REVIEW

**Comment by openjck** [comment]
This works, but it's pretty slow. Still looking into it...

**Comment by openjck** [comment]
Updated and ready for review

**Comment by openjck** [comment]
This won't do much on charts with few data points; I recommend using prod sample data and testing the _bookmarksAtMostRecentDay_ chart.

File: distributionviewer/core/static/js/app/components/views/chart-detail.js [hunk 1]
```diff
@@ -4,10 +4,21 @@ import ChartContainer from '../containers/chart-container';
 
 
 export class ChartDetail extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {showOutliers: false};
```
Comment (spasovski) [inline distributionviewer/core/static/js/app/components/views/chart-detail.js:9]: I agree with this as long as it's obvious enough in the UI.

File: distributionviewer/core/static/js/app/components/views/chart.js [hunk 2]
```diff
@@ -102,10 +50,102 @@ export class Chart extends React.Component {
...
+    /* eslint-disable camelcase */
+    if (showOutliers) {
+      this.setState({
+        min_x: this.pointsMeta[0].x,
+        max_x: this.pointsMeta[pointsMetaLength - 1].x,
+      });
+    } else {
+      this.setState({
+        min_x: this.pointsMeta[Math.max(Math.round(pointsMetaLength * .005) - 1, 0)].x,
+        max_x: this.pointsMeta[Math.min(Math.round(pointsMetaLength * .995) - 1, pointsMetaLength)].x,
```
Comment (spasovski) [inline distributionviewer/core/static/js/app/components/views/chart.js]: The math here is fishy for sure. My understanding is we'll need to figure out the standard deviation and decide to drop any values that exceed it (though technically our request is to use the central 99%). Once we pop those from the array we wouldn't need to mess with min_x, max_x etc.

If the idea really is to just use the "central 99%" we'll need to sort the values in the data (only if it exceeds 99 entries) and then drop the head and tail ends as needed. I can recommend a brand of bourbon that might help with these functions.

**Comment by spasovski** [comment]
I like the component changes but we should rethink the math as well as whether we want to keep regenerating the whole chart. The option should also not be available for any data that doesn't exceed some length threshold like 100.

Honestly this is not a simple task for the project and it might be a good idea to go over this with product to see whether it's even worth it. I'd rather they pick a standard deviation threshold and have us drop any data points that surpass it.

**Comment by openjck** [comment]
Updated the math. Still thinking about the best way to only show the _Show outliers_ option for non-category charts. The chart detail page doesn't know what kind of chart it's going to show, and it would be wasteful to do another request just to find out.

File: distributionviewer/core/static/js/app/components/views/chart-detail.js (hunk 1 above)
Comment (openjck) [reply distributionviewer/core/static/js/app/components/views/chart-detail.js:9]: Sure thing. This is also the default in Brendan's dashboard.

**Comment by spasovski** [comment]
r+pc (I'd like to see product's opinion on this feature first). Thanks for the refactoring work too looks good.

File: distributionviewer/core/static/js/app/components/views/chart.js [hunk 3]
```diff
@@ -102,10 +50,101 @@ export class Chart extends React.Component {
       return chart;
     }
   }
+
+  setMinMaxState(showOutliers) {
+    const pointsMetaLength = this.pointsMeta.length;
+    let minIndex, maxIndex;
+
+    if (showOutliers) {
```
Comment (spasovski) [inline distributionviewer/core/static/js/app/components/views/chart.js]: The math in the `else` would account for it but we could avoid it for smaller data sets by doing `|| pointsMetaLength < 100`. Treat this as a nitpick - it's not necessary.

**Comment by spasovski** [comment]
The product folks seem happy with this option as-is so merge when ready.

**Comment by openjck** [comment]
> The product folks seem happy with this option as-is so merge when ready.

Sure thing. I'll just disable the toggle on non-applicable charts first.

**Comment by openjck** [comment]
Updated and rebased! The checkbox isn't the prettiest thing ever (I prefer the toggle you made for SHIELDDash) but we can follow up with that. The supporting code changes are big enough already.

To test, I recommend using prod endpoints and checking out the _bookmarksAtMostRecentDay_ chart.

File: distributionviewer/core/static/js/app/components/containers/chart-detail-container.js [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
+      <ChartDetail {...this.props}  />
+    );
+  }
+}
 
-export default function(props) {
-  return (
-    <ChartDetail {...props}  />
-  );
+const mapStateToProps = function(store) {
```
Comment (spasovski) [inline distributionviewer/core/static/js/app/components/containers/chart-detail-container.js:14]: You're right we should've done this earlier. Good call.

**Comment by spasovski** [comment]
Looks good but we lost the old prod data. r+ in any case thanks for handling this.

**Pull request merged**